kubeadm: add task page for cluster reconf (#32764)
kubernetes/website: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-reconf.md

---
reviewers:
- sig-cluster-lifecycle
title: Reconfiguring a kubeadm cluster
content_type: task
weight: 10
---

<!-- overview -->

kubeadm does not support automated ways of reconfiguring components that
were deployed on managed nodes. One way of automating this would be
by using a custom [operator](/docs/concepts/extend-kubernetes/operator/).

To modify the components configuration you must manually edit associated cluster
objects and files on disk.

This guide shows the correct sequence of steps that need to be performed
to achieve kubeadm cluster reconfiguration.

## {{% heading "prerequisites" %}}

> 

Reviewer: You need a cluster that was deployed using `kubeadm`.

<!-- steps -->

## Reconfiguring the cluster

kubeadm writes a set of cluster wide component configuration options in
ConfigMaps and other objects. These objects must be manually edited. The command `kubectl edit`
can be used for that.

The `kubectl edit` command will open a text editor where you can edit and save the object directly.

You can use the environment variables `KUBECONFIG` and `KUBE_EDITOR` to specify the location of
the kubectl consumed kubeconfig file and preferred text editor.

For example:
```
KUBECONFIG=/etc/kubernetes/admin.conf KUBE_EDITOR=nano kubectl edit <parameters>
```

{{< note >}}
Upon saving any changes to these cluster objects, components running on nodes may not be
automatically updated. The steps below instruct you on how to perform that manually.
{{< /note >}}

{{< warning >}}
Component configuration in ConfigMaps is stored as unstructured data (YAML string).
This means that validation will not be performed upon updating the contents of a ConfigMap.
You have to be careful to follow the documented API format for a particular
component configuration and avoid introducing typos and YAML indentation mistakes.
{{< /warning >}}

> ### Applying `ClusterConfiguration` changes

Reviewer 2: ```suggestion
### Applying cluster configuration changes
```

Below we use `### Applying kubelet configuration changes`.

Before proceeding with these changes make sure you backup the directory `/etc/kubernetes/`.

To write new certificates you can use:
```shell
kubeadm init phase certs <component-name> --config <config-file>
```

To write new manifest files in `/etc/kubernetes/manifests` you can use:

```shell
kubeadm init phase control-plane <component-name> --config <config-file>
```

The `<config-file>` contents must match the updated `ClusterConfiguration`.
The `<component-name>` value must be the name of the component.

{{< note >}}
Updating a file in `/etc/kubernetes/manifests` will tell the kubelet to restart the component StaticPod.
Try doing these changes one node at a time to leave the cluster without downtime.
{{< /note >}}

### Applying kubelet configuration changes

#### Updating the `KubeletConfiguration`

During cluster creation and upgrade, kubeadm writes its
[`KubeletConfiguration`](/docs/reference/config-api/kubelet-config.v1beta1/)
in a ConfigMap in the `kube-system` namespace called `kubelet-config`.

You can edit the ConfigMap with this command:

```shell
kubectl edit cm -n kube-system kubelet-config
```

The configuration will be located under the `data.kubelet` key.

> #### Reflecting the kubelet changes

Reviewer 2: Can we just run `kubeadm upgrade node` after we edit the configmap? Is there any side-effect?

Author: i need to check.

Author: ok, so we can replace:
> - Edit the contents of the file `/var/lib/kubelet/config.conf` file to
match the contents of the `kubelet-config` ConfigMap

with:
> - Run `kubeadm upgrade node phase kubelet-config` to download the latest `kubelet-config` ConfigMap
> contents into the local file `/var/lib/kubelet/config.conf`

but we can keep the extra step for `/var/lib/kubelet/kubeadm-flags.env`, since it's not managed by `kubeadm upgrade` (unless a particular release changes something there - e.g. flag removal).

kubectl edit service -n kube-system kube-dns
```

#### Reflecting the CoreDNS changes

Once the CoreDNS changes are applied you can delete the CoreDNS Pods:

Obtain the Pod names:

```shell
kubectl get po -n kube-system | grep coredns
```

Delete a Pod with:

```shell
kubectl delete po -n kube-system <pod-name>
```

New Pods with the updated CoreDNS configuration will be created.

{{< note >}}
kubeadm does not allow CoreDNS configuration during cluster creation and upgrade.
This means that if you execute `kubeadm upgrade apply` your changes to the CoreDNS
objects will be lost and must be reapplied.
{{< /note >}}

## Persisting the reconfiguration

During the execution of `kubeadm upgrade` on a managed host, kubeadm might overwrite configuration
that was applied after the cluster was created (reconfiguration).

### Persisting Node object reconfiguration

kubeadm writes Labels, Taints, CRI socket and other information on the Node object for a particular
Kubernetes node. To change any of the contents of this Node object you can use:

```shell
kubectl edit no <node-name>
```

During `kubeadm upgrade` the contents of such a Node might get overwritten.

> If you would like to persist your modifications to the Node object after upgrade

Reviewer 3: ```suggestion
If you would like to persist your modifications to the Node object after upgrade,
```

you can prepare a [kubectl patch](/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/)

> and apply it to the Node object post upgrade:

Reviewer 3: ```suggestion
and apply it to the Node object:
```

```shell
kubectl patch no <node-name> --patch-file <patch-file>
```

#### Persisting control plane component reconfiguration

The main source of control plane configuration is the `ClusterConfiguration`

> object stored in the cluster. To extend the StaticPod manifests configuration

Reviewer 3: ```suggestion
object stored in the cluster. To extend the static Pod manifests configuration,
```

[patches](/docs/setup/production-environment/tools/kubeadm/control-plane-flags/#patches) can be used.

> These patch files must remain as files on the control plane host to ensure that

Reviewer 3: ```suggestion
These patch files must remain as files on the control plane nodes to ensure that
```

they can be used by the `kubeadm upgrade ... --patches <directory>`.

> If reconfiguration is done to the `ClusterConfiguration` stored in the cluster and StaticPod manifests on disk,

Reviewer 3: ```suggestion
If reconfiguration is done to the `ClusterConfiguration` and static Pod manifests on disk,
```
